diff --git a/06_gpu_and_ml/stable_diffusion/playground.py b/06_gpu_and_ml/stable_diffusion/playground.py
index f2819c1..4de30dd 100644
--- a/06_gpu_and_ml/stable_diffusion/playground.py
+++ b/06_gpu_and_ml/stable_diffusion/playground.py
@@ -77,7 +77,7 @@ class Model:
 
 frontend_path = Path(__file__).parent / "frontend"
 
-web_image = modal.Image.debian_slim().pip_install("jinja2")
+web_image = modal.Image.debian_slim().pip_install("jinja2", "fastapi")
 
 
 @stub.function(
@@ -87,27 +87,23 @@ web_image = modal.Image.debian_slim().pip_install("jinja2")
 )
 @modal.asgi_app()
 def app():
-    import fastapi.staticfiles
-    from fastapi import FastAPI
-    from jinja2 import Template
+    from fastapi import FastAPI, Request
+    from fastapi.templating import Jinja2Templates
 
     web_app = FastAPI()
+    templates = Jinja2Templates(directory="/assets")
 
-    with open("/assets/index.html", "r") as f:
-        template_html = f.read()
-
-    template = Template(template_html)
-
-    with open("/assets/index.html", "w") as f:
-        html = template.render(
-            inference_url=Model.web_inference.web_url,
-            model_name="Playground 2.5",
-            default_prompt="Astronaut in the ocean, cold color palette, muted colors, detailed, 8k",
-        )
-        f.write(html)
+    @web_app.get("/")
+    async def read_root(request: Request):
+        return templates.TemplateResponse("index.html", {
+            "request": request,
+            "inference_url": Model.web_inference.web_url,
+            "model_name": "Playground 2.5",
+            "default_prompt": "Astronaut in the ocean, cold color palette, muted colors, detailed, 8k"
+        })
 
     web_app.mount(
-        "/", fastapi.staticfiles.StaticFiles(directory="/assets", html=True)
+        "/static", fastapi.staticfiles.StaticFiles(directory="/assets"), name="static"
     )
 
     return web_app
