diff --git a/06_gpu_and_ml/stable_diffusion/stable_diffusion_xl_lightning.py b/06_gpu_and_ml/stable_diffusion/stable_diffusion_xl_lightning.py
index ffb2f2d..aebf099 100644
--- a/06_gpu_and_ml/stable_diffusion/stable_diffusion_xl_lightning.py
+++ b/06_gpu_and_ml/stable_diffusion/stable_diffusion_xl_lightning.py
@@ -109,7 +109,7 @@ def main(
 
 frontend_path = Path(__file__).parent / "frontend"
 
-web_image = modal.Image.debian_slim().pip_install("jinja2")
+web_image = modal.Image.debian_slim().pip_install("jinja2", "fastapi")
 
 
 @stub.function(
@@ -119,27 +119,24 @@ web_image = modal.Image.debian_slim().pip_install("jinja2")
 )
 @modal.asgi_app()
 def app():
+    from fastapi import FastAPI, Request
+    from fastapi.templating import Jinja2Templates
     import fastapi.staticfiles
-    from fastapi import FastAPI
-    from jinja2 import Template
 
     web_app = FastAPI()
+    templates = Jinja2Templates(directory="/assets")
 
-    with open("/assets/index.html", "r") as f:
-        template_html = f.read()
-
-    template = Template(template_html)
-
-    with open("/assets/index.html", "w") as f:
-        html = template.render(
-            inference_url=Model.web_inference.web_url,
-            model_name="Stable Diffusion XL Lightning",
-            default_prompt="A cinematic shot of a baby raccoon wearing an intricate Italian priest robe.",
-        )
-        f.write(html)
+    @web_app.get("/")
+    async def read_root(request: Request):
+        return templates.TemplateResponse("index.html", {
+            "request": request,
+            "inference_url": Model.web_inference.web_url,
+            "model_name": "Stable Diffusion XL Lightning",
+            "default_prompt": "A cinematic shot of a baby raccoon wearing an intricate Italian priest robe."
+        })
 
     web_app.mount(
-        "/", fastapi.staticfiles.StaticFiles(directory="/assets", html=True)
+        "/static", fastapi.staticfiles.StaticFiles(directory="/assets"), name="static"
     )
 
     return web_app
